<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Homology searching · KmerGMA.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="KmerGMA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">KmerGMA.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Homology searching</a></li><li><a class="tocitem" href="../exactMatch/">Exact gene finding</a></li><li><a class="tocitem" href="../Utils/">Utilities</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Homology searching</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Homology searching</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Qile0317/KmerGMA.jl/blob/master/docs/src/findGenes.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Homology-searching"><a class="docs-heading-anchor" href="#Homology-searching">Homology searching</a><a id="Homology-searching-1"></a><a class="docs-heading-anchor-permalink" href="#Homology-searching" title="Permalink"></a></h1><p>The main function of the package. Based on a novel kmer-based algorithm to reduce runtime over a whole genome to be <em>O(#bp)</em>.</p><p>The algorithm intakes a set of similar reference sequences - for example a collection of V-genes - and iterates over a genome assembly(ies) to find genes that are homologous to the references. </p><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.findGenes" href="#KmerGMA.findGenes"><code>KmerGMA.findGenes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">  findGenes(; genome,
              ref,
              hasN::Bool = true,
              mode::String = &quot;return&quot;,
              fileloc::String = &quot;noPath&quot;,
              k::Int64 = 6,
              windowsize::Int64 = 0,
              thr::Int64 = 0,
              buffer::Int64 = 50,
              resultVec::Vector{FASTA.Record} = FASTA.Record[])</code></pre><p>The main API to find all gene matches in a genome from a reference sequence/reference sequence set. Returns the approximate matches as FASTA records. The descriptions of the record contain information about the match.  The format of the description appears as so:</p><p>&quot;Identifier | SED = a | Pos = b:c | thr = d | buffer = e&quot;</p><p>Where <code>SED</code> indicates how similar the match is to the references, and the lower, the more similar in terms of kmer distance.</p><p>...</p><p><strong>Arguments</strong></p><ul><li><code>genome</code>: Either a String or a Vector of strings, where each string is the file location of a <code>.fasta</code> file containing the genome.</li><li><code>ref</code>: the file location of a fasta file containing the reference sequence(s). The references should be very similar in length. OR a pre-generated reference kmer frequency dictionary.</li></ul><p><strong>Optional Arguments</strong></p><ul><li><code>hasN::Bool = true</code>: Whether both the fasta files from <code>ref</code> and the query <code>genome</code> contains any undefined nucleotide <code>N</code>. If <code>false</code> it runs an alternate algorithm that is alot faster than the one accounting for <code>N</code> nts. </li><li><code>mode = &quot;return&quot;</code>: what the function should return the gene matches in. There are three modes, <code>return</code>, <code>print</code>, <code>write</code>. </li></ul><p>Return mode returns a vector of FASTA records. Print mode simply prints the fasta sequences in the REPL.  Write mode writes the fasta sequences to a fasta file, of which the location has to be given in the <code>fileloc</code> argument</p><ul><li><code>fileloc::String = &quot;noPath&quot;</code>: If the mode argument is <code>&quot;write&quot;</code>, this should be the location of a fasta file that the results should be written into.</li><li><code>k::Int64 = 6</code>: the kmer length to use for approximate matching. Generally it should probably remain between 5 - 10 and probably does not have a profound impact on how the matches are found. Check out the pre-print for more information</li><li><code>windowsize::Int64 = 0</code>: the size of the returned matches. It should be the average length of all reference sequences and is computed automatically if the argument is left as 0.</li><li><code>thr::Int64 = 0</code>: the squared euclidian distance threshold for sequences. Out of the context of the algorithm, lower values mean matches have to be more similar to the references. If left as 0, it is automatically computed. Once again the pre-print has information on this argument.</li><li><code>buffer::Int64 = 0</code>: the amount of nucleotides left and right of a matched sequence that should be added to the returned fasta sequences</li></ul><p>...</p><p>It is recommended to leave all other optional arguments as is, especially the buffer = 50 argument. The purpose of the algorithm is to find approximate matches that can then be BLASTed and aligned.</p><p>However, playing with the <code>thr</code> argument could return more or less matches every time. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/c5dcf8bab574989a9a8d66e283bc70548b21aed9/src/API.jl#L1-L41">source</a></section></article><p>As mentioned in the docstring, if there surely are no &quot;N&quot; nucleotides present, setting <code>HasN</code> to <code>false</code> runs an alternate, much faster algorithm.</p><p>The <code>thr</code> argument can be toyed around with, as increasing the kmer distance threshold doesn&#39;t nessecarily always increase the amount of matches. The default distance is approximate against random sequences and may not always be the best metric.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The unoptimized algorithm that accounts for N nucleotides runs in just under 5 minutes to iterate over a genome of almost 4 billion bps. More benchmarking is on the way. More information is upcoming in a pre-print.</p></div></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../exactMatch/">Exact gene finding »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Friday 9 December 2022 20:09">Friday 9 December 2022</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
