<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Homology searching · KmerGMA.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="KmerGMA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">KmerGMA.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Homology searching</a></li><li><a class="tocitem" href="../exactMatch/">Exact gene finding</a></li><li><a class="tocitem" href="../utilities/">Utilities</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Homology searching</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Homology searching</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Qile0317/KmerGMA.jl/blob/master/docs/src/findGenes.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Homology-searching"><a class="docs-heading-anchor" href="#Homology-searching">Homology searching</a><a id="Homology-searching-1"></a><a class="docs-heading-anchor-permalink" href="#Homology-searching" title="Permalink"></a></h1><p>Here is the main function of the package. Based on a novel kmer-based algorithm to reduce runtime over a whole genome to be <em>O(#bp)</em> on average.</p><p>The algorithm intakes a set of similar reference sequences - for example a collection of V-genes - and iterates over a genome assembly(ies) to find genes that are homologous to the references. </p><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.findGenes" href="#KmerGMA.findGenes"><code>KmerGMA.findGenes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">KmerGMA.findGenes(;
    genome_path::String,
    ref_path::String,
    do_cluster::Bool = false, # to be implemented
    k::Int = 6,
    KmerDistThr::Union{Int64, Float64} = 0,
    buffer::Int64 = 50,
    do_align::Bool = true,
    do_return_dists::Bool = false,
    do_return_hit_loci::Bool = false,
    do_return_align::Bool = false
    verbose::Bool = true)</code></pre><p>The main API to conduct homology searching in a genome, using a kmer-based sequence similarity metric, against a a reference sequence set. For example, the set of all germline V genes of a mammal. Returns the approximate matches as FASTA record vector WITHIN A VECTOR of length 1 in the default configuration of the parameters. The descriptions of the record contain information about the match.  The format of the description appears as so:</p><p>&quot;Identifier | dist = a | MatchPos = b:c | GenomePos = e&quot;</p><p>Where <code>Identifier</code> is the contig ID of the genome where the current hit was found. <code>dist</code> is the kmer similarity, and <code>MatchPos</code> is the unitrange for the match in the contig. <code>GenomePos</code> is the cumulative nucleotides that have been iterated over until the current contig.</p><p>...</p><p><strong>Arguments</strong></p><ul><li><code>genome_path</code>: a string that should be the path of the genome to conduct homology searching on.</li><li><code>ref_path</code>: the file location of a fasta file containing the reference sequence(s). The references should be very similar in length.</li></ul><p><strong>Optional Arguments</strong></p><ul><li><code>do_cluster::Bool = false</code>: work in progress, has no functionality at the moment</li><li><code>k::Int64 = 6</code>: the kmer length to use for approximate matching. Generally it should probably remain between 5 - 10 and probably does not have a profound impact on how the matches are found. Check out the pre-print for more information</li><li><code>KmerDistThr::Int64 = 0</code>: the Kmer Distance distance threshold for sequence matches to the query reference sequence set. Out of the context of the algorithm, lower values mean matches have to be more similar to the references. If left as 0, it is automatically computed. Once again the pre-print has information on this argument.</li><li><code>buffer::Int64 = 50</code>: the amount of nucleotides left and right of a matched sequence that should be added to the returned fasta sequences, as KmerGMA is a heuristic</li><li><code>verbose::Bool = true</code> Indicates whether to show info in the REPL about the the progress of the processing</li><li><code>do_align</code>: Whether to align the hits+bufer region to the consensus sequence of the references. Highly recommended to keep as <code>true</code></li><li><code>do_return_dists</code>: boolean to indicate whether the kmer distances along every window along the genome should be returned in a vector. (intensive memory consumption when genomes are large)</li><li><code>do_return_hit_loci</code>: if <code>true</code>, will return an additional vector of the position within the genomic sequences of each hit, corresponding to the index in the hit vector.</li><li><code>do_return_align</code>: if <code>true</code>, will return an additional vector of alignment object of each hit to the consensus reference sequence.</li><li><code>KmerDist_threshold_buffer::Real = 8.0</code>: a value to determine approximately how much kmer distance a hit should be lower than any random non-hit sequence. Keep in mind that kmerdistance approximates edit distance for mutations better than indels.</li></ul><p>...</p><p>The last three arguments would add term to the output. The output vector would incorporate the respective vectors in the same order of priority if any of the parameters are true.</p><p>Note: Playing with the <code>KmerDistThr</code> argument could return more or less matches every time. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/b21d3683b3c82b08a326b5053df3846744beb888/src/API.jl#L7-L51">source</a></section></article><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The documentation is unfinished and may be somewhat unclear. For more details on how the algorithm functions, some pre-print of the homology searching approach will be uploaded soon.</p></div></div><p>An alternative version of <code>KmerGMA.findGenes</code> which clusters the reference sequence set into similar subsets is also avaliable, though its running speed would be &quot;the number of subsets&quot; times slower. However, the following function is recommended for when speed is less important than accuracy.</p><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.findGenes_cluster_mode" href="#KmerGMA.findGenes_cluster_mode"><code>KmerGMA.findGenes_cluster_mode</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">KmerGMA.findGenes_cluster_mode(;
    genome_path::String,
    ref_path::String,
    cluster_cutoffs = [7,12,20,25],
    k::Int = 6, 
    KmerDistThrs = Float64[0], 
    buff::Int64 = 50,
    do_align::Bool = true,
    do_return_dists::Bool = false,
    do_return_hit_loci::Bool = false,
    do_return_align::Bool = false,
    verbose::Bool = true,
    KmerDist_threshold_buffer::Real = 8.0
    )</code></pre><p>A slower (<code>O(mn)</code>) alternative to <code>KmerGMA.findGenes</code> to conduct homology searching in a genome, against a a reference sequence set, using a kmer-based sequence similarity metric. For example, the set of all germline V genes of a mammal.</p><p>Returns the approximate matches as FASTA record vector WITHIN A VECTOR of length 1 in the default configuration of the parameters. The descriptions of the record contain information about the match.  The format of the description appears as so:</p><pre><code class="nohighlight hljs">&quot;Identifier | dist = a | KFV = b | MatchPos = c:d | GenomePos = e | Len = f&quot;</code></pre><p>Where <code>Identifier</code> is the contig ID of the genome where the current hit was found. <code>dist</code> is the kmer similarity, <code>KFV</code> is the reference kmer frequency vector that a hit was matched against (more info below), <code>MatchPos</code> is the unitrange for the match in the contig. <code>GenomePos</code> is the cumulative nucleotides that have been iterated over until the current contig, and <code>Len</code> is simply the length of the hit sequence</p><p>...</p><p><strong>Arguments</strong></p><ul><li><code>genome_path</code>: a string that should be the path of the genome to conduct homology searching on.</li><li><code>ref_path</code>: the file location of a fasta file containing the reference sequence(s). The references should be very similar in length.</li></ul><p><strong>Optional Arguments</strong></p><ul><li><code>cluster_cutoffs = [7,12,20,25]</code>: Important determination of the cutoff points to construct subsets of similar reference sequences. The cutoff points refer to the kmer distance against the reference. Usually, the more cutoffs may be better but the default cutoffs have been determined to be relatively robust.</li><li><code>k::Int64 = 6</code>: the kmer length to use for approximate matching. Generally it should probably remain between 5 - 10 and probably does not have a profound impact on how the matches are found. Check out the pre-print for more information</li><li><code>KmerDistThrs = [0]</code>: the Kmer Distance distance thresholds for each sequence matche to the query reference sequence set. Out of the context of the algorithm, lower values mean matches have to be more similar to the references. If left as 0, it is automatically computed. Once again a pre-print has information on this argument.</li><li><code>buffer::Int64 = 50</code>: the amount of nucleotides left and right of a matched sequence that should be added to the returned fasta sequences, as KmerGMA is a heuristic</li><li><code>do_align</code>: Whether to align the hits+bufer region to the consensus sequence of the references. Highly recommended to keep as <code>true</code></li><li><code>do_return_dists</code>: boolean to indicate whether the kmer distances along every window along the genome should be returned in a vector. (intensive memory consumption when genomes are large)</li><li><code>do_return_hit_loci</code>: if <code>true</code>, will return an additional vector of the position within the genomic sequences of each hit, corresponding to the index in the hit vector.</li><li><code>do_return_align</code>: if <code>true</code>, will return an additional vector of alignment object of each hit to the consensus reference sequence.</li><li><code>verbose::Bool = true</code> Indicates whether to show info in the REPL about the the progress of the processing</li><li><code>KmerDist_threshold_buffer::Real = 8.0</code>: a value to determine approximately how much kmer distance a hit should be lower than any random non-hit sequence. Keep in mind that kmerdistance approximates edit distance for mutations better than indels.</li></ul><p>...</p><p>The last three arguments would add terms to the output. When <code>verbose</code> is true the exact outputs are even stated.  The output vector would incorporate the respective vectors in the same order of priority if any of the parameters are true.</p><p>There are some more details to be added in future releases.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/b21d3683b3c82b08a326b5053df3846744beb888/src/API.jl#L99-L149">source</a></section></article><p>To write resulting vector of hits into a fasta file, use the following function</p><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.write_results" href="#KmerGMA.write_results"><code>KmerGMA.write_results</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">write_results(KmerGMA_result_vec::Vector{FASTX.FASTA.Record}, file_path::String, width::Int64 = 95)</code></pre><p>Writes all FASTA Records (from the FASTX package) from a vector into a fasta file location <code>file_path</code>. <code>width</code> is an optional argument that indicates the maximum width of sequences written to the file per line.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/b21d3683b3c82b08a326b5053df3846744beb888/src/API.jl#L207-L212">source</a></section></article><p>There is also possible pre-processing of the set of reference sequences that can be used in <code>kmerGMA.findGenes_cluster_mode</code> to further improve accuracy (though usually at the cost of a bit more more speed) More on this will be expanded on.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../exactMatch/">Exact gene finding »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Saturday 18 March 2023 22:32">Saturday 18 March 2023</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
