<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Utilities · KmerGMA.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="KmerGMA.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">KmerGMA.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../findGenes/">Homology searching</a></li><li><a class="tocitem" href="../exactMatch/">Exact gene finding</a></li><li class="is-active"><a class="tocitem" href>Utilities</a><ul class="internal"><li><a class="tocitem" href="#fasta-file-utilities"><span>fasta file utilities</span></a></li><li><a class="tocitem" href="#kmer-utilities"><span>kmer utilities</span></a></li><li><a class="tocitem" href="#paired-kmer-utilities"><span>paired kmer utilities</span></a></li><li><a class="tocitem" href="#strobemer-utilities-(experimental)"><span>strobemer utilities (experimental)</span></a></li><li><a class="tocitem" href="#sequence-divergence-(work-in-progress)"><span>sequence divergence (work in progress)</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Utilities</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Utilities</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Qile0317/KmerGMA.jl/blob/master/docs/src/utilities.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Utility-functions-for-processing-sequences-and-subsequences"><a class="docs-heading-anchor" href="#Utility-functions-for-processing-sequences-and-subsequences">Utility functions for processing sequences and subsequences</a><a id="Utility-functions-for-processing-sequences-and-subsequences-1"></a><a class="docs-heading-anchor-permalink" href="#Utility-functions-for-processing-sequences-and-subsequences" title="Permalink"></a></h1><h2 id="fasta-file-utilities"><a class="docs-heading-anchor" href="#fasta-file-utilities">fasta file utilities</a><a id="fasta-file-utilities-1"></a><a class="docs-heading-anchor-permalink" href="#fasta-file-utilities" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.fasta_id_to_cumulative_len_dict" href="#KmerGMA.fasta_id_to_cumulative_len_dict"><code>KmerGMA.fasta_id_to_cumulative_len_dict</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fasta_id_to_cumulative_len_dict(fasta_file_path::String)</code></pre><p>function to record and store cumulative lengths of the BEGINNING of each record in a dictionary, from a fasta file. for example: for the following fasta file</p><pre><code class="nohighlight hljs">&gt;firstseq
ATGC
&gt;secondseq
AT</code></pre><p>The dictionary returned would be:</p><pre><code class="nohighlight hljs">&quot;firstseq&quot;  =&gt; 4
&quot;secondseq&quot; =&gt; 6</code></pre><p>Not nessecarily in that order.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/ExactMatch.jl#L128-L145">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.getSeq" href="#KmerGMA.getSeq"><code>KmerGMA.getSeq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">getSeq(sequence::FASTA.Record)</code></pre><p>Alias to get the dna longsequence of a fasta record</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/Consts.jl#L28-L32">source</a></section></article><h2 id="kmer-utilities"><a class="docs-heading-anchor" href="#kmer-utilities">kmer utilities</a><a id="kmer-utilities-1"></a><a class="docs-heading-anchor-permalink" href="#kmer-utilities" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.kmer_count" href="#KmerGMA.kmer_count"><code>KmerGMA.kmer_count</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">kmer_count(str::DnaSeq, k::Int, Nt_bits::DnaBits = NUCLEOTIDE_BITS)</code></pre><p>Simple kmer counting function for a DNA BioSequence <code>str</code>, where <code>k</code> is the kmer length to count. The function returns a kmer frequency vector where each INDEX of the vector corresponds to a unique kmer. For each index, the unsigned bits of the index correspond to a binary representation of sequences where two bits encode one nucleotide. To see what kmer each index corresponds to, see the function <code>as_kmer</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/Kmers.jl#L4-L11">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.as_kmer" href="#KmerGMA.as_kmer"><code>KmerGMA.as_kmer</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">as_kmer(kmer_uint::Integer, kmer_len::Int,Nt_bits::Dict{UInt, Seq} = BitNtDict)</code></pre><p>Takes a positive integer representing a 2-bit-based dna kmer, and the actual length of the kmer, and returns the BioSequences dna sequence of the kmer.  The Nt_bits argument can be ignored. Additionally, note that both input parameters will be modified to 0 or 1.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/Kmers.jl#L73-L78">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.as_UInt" href="#KmerGMA.as_UInt"><code>KmerGMA.as_UInt</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">UInt(kmer_seq::LongSequence{DNAAlphabet{4}}, Nt_bits::DnaBits = NUCLEOTIDE_BITS)</code></pre><p>Convert a dna biosequence <code>kmer_seq</code> into an unsigned integer. Users can ignore the second argument</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/Kmers.jl#L95-L99">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.kmer_dist" href="#KmerGMA.kmer_dist"><code>KmerGMA.kmer_dist</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">kmer_dist(seq1, seq2, k::Int, Nt_bits::DnaBits = NUCLEOTIDE_BITS)</code></pre><p>returns the kmer distance between two biosequences <code>seq1</code> and <code>seq2</code>, where <code>k</code> is the kmer length. Users can ignore the last argument.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/Kmers.jl#L47-L52">source</a></section></article><h2 id="paired-kmer-utilities"><a class="docs-heading-anchor" href="#paired-kmer-utilities">paired kmer utilities</a><a id="paired-kmer-utilities-1"></a><a class="docs-heading-anchor-permalink" href="#paired-kmer-utilities" title="Permalink"></a></h2><p>The following function count all kmer pairs within a sequence, which is a feature to be further utilized in an upcoming release. Counting paired kmers could serve as a more robust alternative to regular kmers in many applications.</p><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.kmer_pair_count" href="#KmerGMA.kmer_pair_count"><code>KmerGMA.kmer_pair_count</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">kmer_pair_count(seq::DnaSeq, k::Int = 3, Nt_bits::DnaBits = NUCLEOTIDE_BITS)</code></pre><p>Counts all paired kmers in a Biosequences DNA longsequence (or a longsubseq), where each index of the resulting <code>paired kmer frequency vector</code> correspond to the first and second kmers appended together.</p><p>For example, the 2mer pair <code>AT</code> and <code>GC</code> in a sequence <code>ATGC</code> would have a count of 1 in the resulting vector of length 4^(2*2) = 1 It would be at the index in the vector equivalent to <code>as_UInt(dna&quot;ATGC&quot;)</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/PairedKmers.jl#L27-L35">source</a></section></article><h2 id="strobemer-utilities-(experimental)"><a class="docs-heading-anchor" href="#strobemer-utilities-(experimental)">strobemer utilities (experimental)</a><a id="strobemer-utilities-(experimental)-1"></a><a class="docs-heading-anchor-permalink" href="#strobemer-utilities-(experimental)" title="Permalink"></a></h2><p>Strobemers consist of two or more linked shorter k-mers, where the combination of linked k-mers is decided by a hash function. This implementation is currently being experimented with and is unoptimized, but may be implemented in the near future as a new homology searching algorithm. </p><p>The strobemers implemented are simply randstrobes with two k-mers, and in the future will be a part of a new module <code>StrobemerGMA</code></p><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.get_strobe_2_mer" href="#KmerGMA.get_strobe_2_mer"><code>KmerGMA.get_strobe_2_mer</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">get_strobe_2_mer(
    seq::LongSequence{DNAAlphabet{4}},
    s::Int = 2,
    w_min::Int = 3,
    w_max::Int = 5,
    q::Int = 5;
    withGap::Bool = true)</code></pre><p>Get the randstrobe that consists of two kmers of the current sequence <code>seq</code>.  The <code>withGap</code> named argument indicates whether to return the strobemer with sequence gaps or just as a regular kmer with all gaps removed</p><p>...</p><p><strong>Strobemer parameters</strong></p><ul><li><code>s::Int = 2</code>: length of the kmers that consists of the strobemer.</li><li><code>w_min::Int = 3</code>: the start of the window to obtain the second kmer from.</li><li><code>w_max::Int = 5</code>: the end of the window to obtain the second kmer from.</li><li><code>q::Int = 5</code>: the prime number that the randstrobe hashing function should use. </li></ul><p>...</p><p>It is recommended to keep all optional strobemer parameters as is</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/StrobemerGMA/Strobemers.jl#L16-L38">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.ungapped_strobe_2_mer_count" href="#KmerGMA.ungapped_strobe_2_mer_count"><code>KmerGMA.ungapped_strobe_2_mer_count</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ungapped_strobe_2_mer_count(
    seq::DnaSeq;
    s::Int = 2,
    w_min::Int = 3,
    w_max::Int = 5,
    q::Int = 5)</code></pre><p>Get the ungapped randstrobe (two kmers of the current sequence <code>seq</code>) frequency vector.  The vector is <span>$4^{2s}$</span> long, where each index corresponds to the strobemer with gaps removed. So the strobemer <code>AC--GT--</code> would correspond to the index corresponding to <code>ACGT</code>. To convert from index to the ungapped strobemer, see <code>KmerGMA.as_kmer</code></p><p>...</p><p><strong>Strobemer parameters</strong></p><ul><li><code>s::Int = 2</code>: length of the kmers that consists of the strobemer.</li><li><code>w_min::Int = 3</code>: the start of the window to obtain the second kmer from.</li><li><code>w_max::Int = 5</code>: the end of the window to obtain the second kmer from.</li><li><code>q::Int = 5</code>: the prime number that the randstrobe hashing function should use. </li></ul><p>...</p><p>It is currently the responsibility of the user to ensure that all parameters do not conflict with eachother </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/StrobemerGMA/Strobemers.jl#L61-L83">source</a></section></article><h2 id="sequence-divergence-(work-in-progress)"><a class="docs-heading-anchor" href="#sequence-divergence-(work-in-progress)">sequence divergence (work in progress)</a><a id="sequence-divergence-(work-in-progress)-1"></a><a class="docs-heading-anchor-permalink" href="#sequence-divergence-(work-in-progress)" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.mutate_seq" href="#KmerGMA.mutate_seq"><code>KmerGMA.mutate_seq</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">mutate_seq(seq::LongSequence{DNAAlphabet{4}}, mut_rate::Real)</code></pre><p>returns a new biosequence where <code>mute_rate</code> portion of the input biosequence <code>seq</code> are substituted to a different base pair. The <code>mut_rate</code> should be between 0 and 1.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/DistanceTesting.jl#L57-L62">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="KmerGMA.mutate_seq!" href="#KmerGMA.mutate_seq!"><code>KmerGMA.mutate_seq!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">mutate_seq!(seq::LongSequence{DNAAlphabet{4}}, mut_rate::Real)</code></pre><p>randomly subsitutes <code>mute_rate</code> portion of the input biosequence <code>seq</code> in place. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Qile0317/KmerGMA.jl/blob/8c0243b7baed34e3cac497a96b607725a0fbcf4f/src/DistanceTesting.jl#L44-L48">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../exactMatch/">« Exact gene finding</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Saturday 25 March 2023 17:53">Saturday 25 March 2023</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
